library/x509_create.c

```diff
@@ -86,62 +92,161 @@ static const char *x509_at_oid_from_name( const char *name, size_t name_len )
+    {
+        return( MBEDTLS_ERR_X509_ALLOC_FAILED );
+    }
+
+    /*
+     * Skip white spaces
+     */
+    while( c < end && *(c + 1) == ' ' )
+        c++;
+
+    *s = c + 1;
+
+    return ( 0 );
+}
+
 int mbedtls_x509_string_to_names( mbedtls_asn1_named_data **head, const char *name )
 {
     int ret = 0;
+    /*
+     * s points to the current string name.
+     * c points to the current character in the string being parsed.
+     * d points to the end of the current parsed named data.
+     */
     const char *s = name, *c = s;
     const char *end = s + strlen( s );
     const char *oid = NULL;
-    int in_tag = 1;
     char data[MBEDTLS_X509_MAX_DN_NAME_SIZE];
     char *d = data;
+    x509_dn_parse_state dn_cur_state = dn_in_tag;
 
     /* Clear existing chain if present */
     mbedtls_asn1_free_named_data_list( head );
 
     while( c <= end )
     {
-        if( in_tag && *c == '=' )
+        switch( dn_cur_state )
         {
-            if( ( oid = x509_at_oid_from_name( s, c - s ) ) == NULL )
+            /*
+             *Currently parsing a DN tag.
+             */
+            case dn_in_tag:
             {
-                ret = MBEDTLS_ERR_X509_UNKNOWN_OID;
-                goto exit;
+                if( *c == '=' )
+                {
+                    if( ( oid = x509_at_oid_from_name( s, c - s ) ) == NULL )
+                    {
+                        ret = MBEDTLS_ERR_X509_UNKNOWN_OID;
+                        goto exit;
+                    }
+                    s = c + 1;
+                    d = data;
+                    if( *s == '\"' )
+                    {
+                        dn_cur_state = dn_in_quoted_value;
+                        s++;
+                        c++;
+                    }
+                    else
+                    {
+                        dn_cur_state = dn_in_unquoted_value;
+                    }
+                }
             }
+            break;
 
-            s = c + 1;
-            in_tag = 0;
-            d = data;
-        }
-
-        if( !in_tag && *c == '\\' && c != end )
-        {
-            c++;
-
-            /* Check for valid escaped characters */
-            if( c == end || *c != ',' )
+            /*
+             * Currently parsing a quoted value.
+             */
+            case dn_in_quoted_value:
             {
-                ret = MBEDTLS_ERR_X509_INVALID_NAME;
-                goto exit;
+                if( *c == '\"' && c != end )
+                {
+                    /*
+                     * Skip white spaces
+                     */
+                    while( c < end && *(c + 1) == ' ' )
+                        c++;
+
+                    c++;
+
+                    if( *c != ',' && c != end )
+                    {
+                        ret = MBEDTLS_ERR_X509_INVALID_NAME;
+                        goto exit;
+                    }
+                    ret = dn_store_named_data( head, oid, data,
```

Comment:
The whitespace skipping is not handled both here and in `dn_store_named_data`. Given that it is also used before the `=` which separates OID and value, I think there should be a uniform whitespace-skipping function which is used in all states, and `dn_store_named_data` should be removed in favor of a direct call to `mbedtls_asn1_store_named_data`.

Reply:
But then we will have some duplication in every state.

Reply:
@RonEld Why? We could have a static function `skip_spaces` or similar which is called whenever potential whitespace needs to be skipped.

Reply:
`dn_stored_named_data` is essentially a white space skip + call to `mbedtls_asn1_store_named_data`.

Reply:
It also checks that `d - data` doesn't exceed the size limit.
Personally, I thought it would be more readablehaving it all ion the same function, but if you object, I'll remove it

Reply:
I object to having it non-uniform (and hence less readable) which it currently is. 

My preference would be to remove `dn_store_named_data()` and potentially introduce a white space skipping function instead. It doesn't matter that it's only a few lines of code - the compiler may inline it if it thinks it's worth it - but it'll nonetheless increase readability.